doc/monads/readers.lean

```diff
@@ -0,0 +1,196 @@
+early `return` in a monadic function similar to how it can in an imperative language while `pure`
+cannot.
+
+> So technically if `return` is the last statement in a function it could be replaced with `pure <|`,
+but one could argue that `return` is still a little easier for most folks to read, just so long as
+you understand that `return` is doing more than other languages, it is also wrapping pure values in
+the monadic container type.
+
+> **Side note 2**: If the function `readerFunc3` also took some explicit arguments then you would have
+to write `(readerFunc3 args).run env` and this is a bit ugly, so Lean provides an infix operator
+`|>` that eliminiates those parens so you can write `readerFunc3 args |>.run env` and then you can
+chain multiple monadic actions like this `m1 args1 |>.run args2 |>.run args3` and this is the
+recommended style.  You will see this patten used heavily in Lean code.
+
+The `let env ← read` expression in `readerFunc1` unwraps the environment from the `ReaderM` so we
+can use it. Each type of monad might provide one or more extra functions like this, functions that
+become available only when you are in the context of that monad.
+
+Here the `readerFunc2` function uses the `bind` operator `>>=` just to show you that there are bind
+operations happening here.  The `readerFunc3` function uses the  `do` notation you learned about in
+[Monads](monads.lean.md) which hides that bind operation and can make the code look cleaner.
+
+The `do` notation with `let x ← readerFunc2` is also calling the `run` function to actually execute
+the monadic action, then it unwraps the value x, so it is really doing `let x ← readerFunc2.run`.
+
+The important difference here to the earlier code is that `readerFunc3` and `readerFunc2` no longer
+have an **explicit** Environment input parameter that needs to be passed along all the way to
+`readerFunc1`.  Instead, the `ReaderM` monad is taking care of that for you, which gives you the
+illusion of something like global state where the state is now available to all functions that use
+the `ReaderM` monad.
+
+The above code also introduces an important idea. Whenever you learn about a monad "X", there's
+often (but not always) a `run` function to execute that monad, and sometimes some additional
+functions like `read` that interact with the monad state.
+
+You might be wondering, how does the state actually move through the `ReaderM` monad? How can you
+add an input argument to a function by modifying it's return type?  There is a special command in
+the Lean interpreter that will show you the reduced Types:
+-/
+#reduce ReaderM Environment String   -- Environment → String
+/-!
+And you can see here that this type is actually a function!  It's a function that takes an
+`Environment` as input and returns a `String`.  So, remember in Lean that a function that takes
+argument a and returns a string: `def f (a : Nat) → String` is the same as the function that takes no
```

Comment:
This paragraph is incorrect. We cannot say that `f : Nat -> String` is a function that takes no arguments and returns another function of type `Nat -> String`.

Reply:
I was trying to say  `def f (a : Nat) → String` can be rewritten as `def f : Nat → String` to help make the connection to the user that this is how ReaderM can add an input argument even though ReaderM is used as a return type.  It is building on Lean's support for "currying" ?  What is a better way to say all this?

Reply:
This is not related to currying. If I remember correctly, David's book explains currying. You can also check the page on Wikipedia. 
The types `ReaderM Environment String` and `Environment -> String` are definitionally equal in Lean. The following definition is type correct in Lean
```lean
def f : ReaderM Environment String :=
  fun (_ : Environment) => "hello world"
``` 
The expression `fun (_ : Environment) => "hello world"` has type `Environment -> String`.
```lean
#check fun (_ : Environment) => "hello world" 
```
Lean accepts the definition because the types are definitionally equal. TPIL mentions definitionally equality several times. I think it is worth keeping re-reading TPIL until the Lean type system is clear to you.

Reply:
Ok, so building on that I can see that these are definitially equal, so I will use that terminology:
```lean
def f (a : Nat) : String := "abc"

#print f -- fun a => "abc"

def g : Nat → String := λ a => "abc"

#print g -- fun a => "abc"

example : f = g := by rfl  -- Goals accomplished 🎉
```
I will not call this `currying` even though that's exactly what David called it here:
https://github.com/leanprover/fp-lean/blob/master/functional-programming-lean/src/getting-to-know/functions-and-definitions.md?plain=1#L70

Reply:
PS: I had a typo in the original text that may explain all your confusion here, I had `def f (a : Nat) → String` instead of `def f (a : Nat) : String`.

I have updated this text in https://github.com/leanprover/lean4/pull/1505/commits/9b24f8de09a70d3793d040aff7d217bed8be8acc, so please me know if the new version is better.

Reply:
> I will not call this currying even though that's exactly what David called it here:
https://github.com/leanprover/fp-lean/blob/master/functional-programming-lean/src/getting-to-know/functions-and-definitions.md?plain=1#L70

Read the text again. You will see it is different. He has a binary function.